bindings/java/src/org/sleuthkit/datamodel/CaseDbAccessManager.java

```diff
@@ -758,18 +835,36 @@ private void validateSQL(String sql) throws TskCoreException {
 		 * @throws SQLException
 		 * @throws TskCoreException
 		 */
-		private CaseDbPreparedStatement(String query, boolean isWriteLockRequired) throws SQLException, TskCoreException {		
+		private CaseDbPreparedStatement(StatementType type, String query, boolean isWriteLockRequired) throws SQLException, TskCoreException {		
 			if (isWriteLockRequired) {
-				CaseDbAccessManager.this.tskDB.acquireSingleUserCaseWriteLock();	
+				CaseDbAccessManager.this.tskDB.acquireSingleUserCaseWriteLock();
+				this.lockType = LockType.WRITE;
 			} else {
-				CaseDbAccessManager.this.tskDB.acquireSingleUserCaseReadLock();	
+				CaseDbAccessManager.this.tskDB.acquireSingleUserCaseReadLock();
+				this.lockType = LockType.READ;
 			}
-			
-			this.hasWriteLock = isWriteLockRequired;
 			this.connection = tskDB.getConnection();
 			this.preparedStatement = connection.getPreparedStatement(query, Statement.NO_GENERATED_KEYS);
 			this.originalSql = query;
-			
+			this.type = type;
+		}
+		
+		/**
+		 * Construct a prepared statement using an already open transaction.
+		 *
+		 * @param type                The type of statement.
+		 * @param query               The query string.
+		 * @param trans               The open transaction.
+		 *
+		 * @throws SQLException
+		 * @throws TskCoreException
+		 */
+		private CaseDbPreparedStatement(StatementType type, String query, CaseDbTransaction trans) throws SQLException, TskCoreException {		
+			this.lockType = LockType.NONE;
+			this.connection = trans.getConnection();
+			this.preparedStatement = connection.getPreparedStatement(query, Statement.NO_GENERATED_KEYS);
```

Comment:
It looks like we are returning a cached prepared statement here from the adHocPreparedStatements without checking if it has been closed already or not. 
```   try (CaseDbAccessManager.CaseDbPreparedStatement statement = caseDbAccessManager.prepareInsert(TABLE_NAME, insertSQL, caseDbTransaction)) {
```
the above code causes NPE when called a second time. 

I think we should not return cached statements that are closed. 
we should also document that it is the caller's responsibility to keep the statement open if they are doing other things with it ? I am not a 100% sure how to handle this part.

Reply:
No we just shouldn't close the prepared statements when we have a transaction. They get closed when the connection is closed. But I also added a check that we aren't supplying closed statements just in case.